Zend/Optimizer/block_pass.c

```diff
@@ -274,7 +274,9 @@ static void zend_optimize_block(zend_basic_block *block, zend_op_array *op_array
 				 * If it's not local, then the other blocks successors must also eventually either FREE or consume the temporary,
 				 * hence removing the temporary is not safe in the general case, especially when other consumers are not FREE.
 				 * A FREE may not be removed without also removing the source's result, because otherwise that would cause a memory leak. */
-				if (opline->op1_type == IS_TMP_VAR) {
+				if (opline->extended_value == ZEND_FREE_VOID_CAST) {
```

Comment:
This is too conservative. The following code will now still have a useless FREE but it previously didn't:
```php
<?php
(void)(bool)foo();
```

Reply:
Is that a problem, though? Writing `(void)` is an intentional choice by the user, it does not happen by accident.

Given that its explicit purpose is to keep the expression alive, I'd say it's working as intended and also don't see how handling that case would enable other (meaningful) optimizations to happen either. Thus it looks like added complexity for little benefit to me.

Reply:
It could miss some optimization opportunities when people add `(void)` in different places, but sure I suppose it doesn't matter that much.
I disagree it's much added complexity though

Reply:
As noted on Slack: `(void)(bool)flock(...)`. OPcache knows that `flock()` can only return a `bool` and if the `(void)` is optimized away, `(bool)` would also be legal to optimize, resulting in `flock(...)` standing on its own and thus reporting the `#[\NoDiscard]` warning that `(void)` is intended to suppress.

The only legal optimization here would be `(void)(bool)flock(...)` -> `(void)flock(...)`, which is what I mean by added complexity.